File: PixelDungeon/src/main/java/com/nyrds/retrodungeon/mobs/guts/TreacherousSpirit.java
package com.nyrds.retrodungeon.mobs.guts;

import android.support.annotation.NonNull;

import com.nyrds.retrodungeon.items.guts.HeartOfDarkness;
import com.watabou.pixeldungeon.Dungeon;
import com.watabou.pixeldungeon.actors.Actor;
import com.watabou.pixeldungeon.actors.Char;
import com.watabou.pixeldungeon.actors.mobs.Mob;
import com.watabou.pixeldungeon.effects.Pushing;
import com.watabou.utils.Random;

/**
 * Created by DeadDie on 12.02.2016
 */
public class TreacherousSpirit extends Mob {
    {
        hp(ht(200));
        defenseSkill = 35;

        exp = 45;
        maxLvl = 30;

        setState(WANDERING);
        lootChance = 1f;
        loot = HeartOfDarkness.class;
    }

    @Override
    public int attackProc(@NonNull Char enemy, int damage ) {
        //Summon proc
        if (Random.Int(4) == 1){
            int spiritPos = Dungeon.level.getEmptyCellNextTo(getPos());

            if (Dungeon.level.cellValid(spiritPos)) {
                SpiritOfPain spirit = new SpiritOfPain();
                spirit.setPos(spiritPos);
                Dungeon.level.spawnMob(spirit, 0);
                Actor.addDelayed(new Pushing(spirit, getPos(), spirit.getPos()), -1);
            }
        }
        return damage;
    }

    @Override
    public int damageRoll() {
        return Random.NormalIntRange(30, 45);
    }

    @Override
    public int attackSkill( Char target ) {
        return 35;
    }

    @Override
    public int dr() {
        return 25;
    }

    @Override
    public boolean canBePet(){
        return false;
    }

}

File: scripts/mobs/TreacherousSpirit.lua
local RPD = require "scripts/lib/commonClasses"
local mob = require "scripts/lib/mob"

return mob.init{
    attackProc = function(self, enemy, dmg)
        -- 1/4 chance to summon a Spirit of Pain nearby
        if enemy ~= nil and math.random(4) == 1 then
            local level = RPD.Dungeon.level
            local pos   = self:emptyCellNextTo()
            if level:cellValid(pos) then
                local spirit = RPD.MobFactory:mobByName("SpiritOfPain")
                spirit:setPos(pos)
                level:spawnMob(spirit)
            end
        end
        return dmg
    end
}
